### ISSUE TITLE
Prohibit non-derived Newtype instances

### ISSUE THREAD

**Issue by rightfold** [purescript/purescript#2895]
I observe quite frequently that I need `map wrap` or `map unwrap`. This is obviously wasteful, since all `Newtype` instances I know of are derived, and so `wrap` and `unwrap` are always no-ops. Unless someone can convince me of useful non-derived `Newtype` instances I'd like to propose a restriction on `Newtype` instances such that they cannot be written without deriving them. That way we get a guarantee that `wrap` and `unwrap` are no-ops. This guarantee can be used to optimize `map wrap` and `map unwrap` to `unsafeCoerce`, either automatically or through additional functions in the purescript-newtype package.

**Comment by rightfold** [comment]
Actually, I think this may cause a safety issue with unlawful `Functor` instances.

```purescript
data F a = X | Y (a ~ Int) a

instance functorF :: Functor F where
  map _ _ = X

bad :: F (Additive Int)
bad = unsafeCoerce (Y id 1)
```

We now have a proof that `Int ~ Additive Int`. It wouldn't happen with the unoptimized `map wrap`.

**Comment by paf31** [comment]
Why not just make it a law? I don't particularly like the idea of type classes whose instances must be derived.

**Comment by rightfold** [comment]
Is it fine to compromise type safety in case of unlawful instances?

**Comment by paf31** [comment]
Well any use of `unsafeCoerce` potentially compromises type safety. You have to rely on something to make sure it's safe, and in this case it's a type class law.

**Comment by rightfold** [comment]
Then I'm fine with that.

**Comment by mat8913** [comment]
Could [Haskell's solution] to this problem also be applied to Purescript?

[Haskell's solution]: http://cs.brynmawr.edu/~rae/papers/2014/coercible/coercible-ext.pdf

**Comment by paf31** [comment]
I don't believe that's the same issue as the one here. The roles stuff deals with newtype deriving for types involving things like type families.

**Comment by mat8913** [comment]
The issue is that you want `map unwrap` and `map wrap` to automatically optimize to `unsafeCoerce`, but only when it's safe, right?

The paper I linked introduces a new typeclass `Coercible` which is a bit like your `Newtype`, but more general. An instance of `Coercible a b` means that it is safe to do coercions between types `a` and `b`. At the moment, you have an instance `Newtype (Additive Int) Int`. Under the paper's proposal, you would not only have instance `Coercible (Additive Int) Int` but also `Coercible (Array (Additive Int)) (Array Int)`. Importantly, you would *not* have instance `Coercible (F (Additive Int)) (F Int)`.

The paper also introduces a function `coerce :: Coercible a b => a -> b`.

Therefore, by rewriting `map coerce` to `coerce` whenever it typechecks, you get the optimization when possible and avoid the safety issue with the unlawful Functor `F`.

**Comment by paf31** [comment]
I'm sorry, I was confusing this issue with the open one about newtype deriving 😄 Thanks for the reference, I'll have a think about it. I dare say we could have the solver generate instances for `Coercible (f a) (f b)` in terms of `Coercible a b`.

It might make an interesting project for someone who wants to get familiar with the type checker code.

**Comment by rightfold** [comment]
This sounds nice indeed. I thought Coercible was more like Newtype, I didn't know it required that the representation be the same

Sent from my iPhone.

**Comment by rightfold** [comment]
Other idea from IRC:

1. Add a `SameRepr :: Type -> Type -> Constraint` class which is always resolved by the compiler (like some other classes we have). An instance for it exists if two types have the same representation. This is the case when one of the following hold:

   - Both are the same type.
   - Either one is a newtype for the other, or indirectly so (transitive).

2. Make `SameRepr a b` a superclass of `Newtype a b`. This guarantees that `unsafeCoerce :: a -> b` and `unsafeCoerce :: b -> a` are safe.

3. Move the `wrap` and `unwrap` functions out of the `Newtype` class, making them specializations of `unsafeCoerce` that have `Newtype` constraints instead.

In short:

 - This would solve the problem.
 - `Newtype a b` becomes an opt-in for a safe no-op isomorphism between `a b`.
 - `SameRepr a b` is not necessarily a safe isomorphism because a newtype may have a smart constructor (in which case there is no `Newtype` instance).

**Comment by matthewleon** [comment]
Has anyone given this proposal any further thought?

**Comment by hdgarrood** [comment]
I think we should just do `Coercible` instead, with roles, especially since we have a PR for it now. See #3302 and #3351.

**Comment by hdgarrood** [comment]
I'm going to close this because we intend to add `Coercible`, which (as far as I can tell) solves the original problem.
